project_task_materials_stock/models/project.py

```diff
@@ -0,0 +1,180 @@
+# -*- coding: utf-8 -*-
+# Copyright 2015 Antiun Ingeniería S.L. - Sergio Teruel
+# Copyright 2015 Antiun Ingeniería S.L. - Carlos Dauden
+# Copyright 2016 Tecnativa - Vicent Cubells
+# License AGPL-3 - See http://www.gnu.org/licenses/agpl-3.0.html
+from openerp import models, fields, api, exceptions, _
+
+
+class ProjectTaskType(models.Model):
+    _inherit = 'project.task.type'
+
+    consume_material = fields.Boolean(
+        string='Consume Material',
+        help="If you mark this check, when a task goes to this state,"
+             "it will consume the associated materials")
+
+
+class Task(models.Model):
+    _inherit = "project.task"
+
+    @api.multi
+    @api.depends('material_ids.stock_move_id')
+    def _compute_stock_move(self):
+        for task in self:
+            task.stock_move_ids = task.mapped('material_ids.stock_move_id')
+
+    @api.multi
+    @api.depends('material_ids.analytic_line_id')
+    def _compute_analytic_line(self):
+        for task in self:
+            task.analytic_line_ids = task.mapped(
+                'material_ids.analytic_line_id')
+
+    @api.multi
+    @api.depends('stock_move_ids.state')
+    def _compute_stock_state(self):
+        for task in self:
+            if not task.stock_move_ids:
+                task.stock_state = 'pending'
+            elif task.stock_move_ids.filtered(
+                    lambda r: r.state == 'confirmed'):
+                task.stock_state = 'confirmed'
+            elif task.stock_move_ids.filtered(lambda r: r.state == 'assigned'):
+                task.stock_state = 'assigned'
+            elif task.stock_move_ids.filtered(lambda r: r.state == 'done'):
+                task.stock_state = 'done'
+
+    stock_move_ids = fields.Many2many(
+        comodel_name='stock.move', compute='_compute_stock_move',
+        string='Stock Moves')
+    analytic_line_ids = fields.Many2many(
+        comodel_name='account.analytic.line', compute='_compute_analytic_line',
+        string='Analytic Lines')
+    consume_material = fields.Boolean(related='stage_id.consume_material')
+    stock_state = fields.Selection(
+        [('pending', 'Pending'),
+         ('confirmed', 'Confirmed'),
+         ('assigned', 'Assigned'),
+         ('done', 'Done')],
+        compute='_compute_stock_state', string='Stock State')
+
+    @api.multi
+    def unlink_stock_move(self):
+        res = False
+        moves = self.mapped('stock_move_ids')
+        moves_done = moves.filtered(lambda r: r.state == 'done')
+        if not moves_done:
+            moves.filtered(lambda r: r.state == 'assigned').do_unreserve()
+            moves.filtered(
+                lambda r: r.state in ['waiting', 'confirmed', 'assigned']
```

Comment:
Use a set instead of a list, [it's way faster in this case](http://stackoverflow.com/questions/2831212/python-sets-vs-lists#2831242).